Set TimePicker.Time seconds to zero when UseSeconds is False

VisualMelon opened the pull request:
<!--- See CONTRIBUTING.md for general guidelines on contributions -->

## What does the pull request do?
Addresses the most significant issue (breaking change) discussed in https://github.com/AvaloniaUI/Avalonia/issues/17024


## What is the current behavior?
The seconds component of the Time property is always set to an uncontrolled value when `UseSeconds` is false.


## What is the updated/expected behavior with this PR?
When `UseSeconds` is false, the Time property has its seconds component set to zero


## How was the solution implemented (if it's not obvious)?
N/A


## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
N/A

## Obsoletions / Deprecations
N/A

## Fixed issues
Fixes part of #17024

MrJul commented:
Thank you, the change looks reasonable to me.
Can you please add a short unit test for this case?

VisualMelon commented:
I had a go at that, and my AV got upset, but I'll try again. May take a bit of time for me to work out how to test it sensibly.

VisualMelon committed:
Set TimePickerPresenter.Time seconds component to zero when UseSeconds is false

VisualMelon commented:
Pushed a couple of simple tests: not sure if they are up to muster; obviously I'm happy to change anything with a bit of direction. Should the `TimePickerPresenter` have its own tests file?

I had a lot of trouble with the templating the `TimePickerPresenter`:
 - Added a couple of `TemplatePart` declarations to it that seemed to be missing (not sure if they're inherited from somewhere else?)
 - Changed the order that the `ItemFormat` and `SelectedValue` are assigned in `InitPicker`, as `SelectedValue` was crashing beforehand
 - I seemed to have to call `pickerPresenter.ApplyTemplate();` inside the `TimePicker` template, but it still seemed to crash inside `ApplyTemplate` when I didn't (before the above fix was put in), so I'm just confused what's going on there

Note that the test explicitly invoke the `AcceptBtn`: the seconds property will be (59/60 times) non-zero before this is pressed. See discussion in #17024; I think this behaviour is the same as 11.1 anyway, so not a breaking change.

VisualMelon commented:
Looks like I've broken something, but no idea why (Appium makes my AV unhappy to didn't run those locally). Will take another look in a few hours/tomorrow.

VisualMelon commented:
Rebased against master and tests stopped failing; not sure those are related.

maxkatz6 commented on src/Avalonia.Controls/DateTimePickers/TimePickerPresenter.cs:
```diff
@@ -28,7 +28,9 @@ namespace Avalonia.Controls
     [TemplatePart("PART_PeriodUpButton",   typeof(RepeatButton))]
     [TemplatePart("PART_PickerContainer",  typeof(Grid), IsRequired = true)]
-    [TemplatePart("PART_ThirdSpacer",     typeof(Rectangle), IsRequired = true)]
+    [TemplatePart("PART_FirstSpacer",      typeof(Rectangle), IsRequired = true)]
+    [TemplatePart("PART_SecondSpacer",     typeof(Rectangle), IsRequired = true)]
+    [TemplatePart("PART_ThirdSpacer",      typeof(Rectangle), IsRequired = true)]
```
Just a sanity check, are these really required? I.e. will control work fine, if these spacers were missing, without any C# code change.

VisualMelon replied:
`ApplyTemplate` crashed when I didn't include them in the test template, so I added them here as they seemed to be missing.

VisualMelon replied:
Looks like FirstSpacer doesn't need to exist (isn't used at all, which is a bit odd), but SecondSpacer is assumed in the code. I'll remove FirstSpacer or make any other suggested changes tomorrow.

MrJul replied:
Yep, they're unfortunately needed currently. We should refactor these templates someday: I'm sure a lot of parts could be made optional and/or simple styles added to the themes instead. Anyways, such a refactoring isn't the point of this PR :)

VisualMelon committed:
Remove unneeded PART_FirstSpace annotation from TimePickerPresenter

MrJul commented:
> Rebased against master and tests stopped failing; not sure those are related.

They aren't related, these specific integration tests are unfortunately flaky, you can ignore them for now. I'm hoping #17259 will help them pass...

MrJul reviewed (approved):
LGTM, thanks!

MrJul commented:
Your branch is write protected, can you please merge/rebase from master to trigger the CI again?

MrJul committed:
Merge branch 'master' into useseconds
